core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -143,9 +154,24 @@ static void prepareHeaders(Metadata headers, CallOptions callOptions, String use
   }
 
   @Override
-  public void start(Listener<RespT> observer, Metadata headers) {
+  public void start(final Listener<RespT> observer, Metadata headers) {
     Preconditions.checkState(stream == null, "Already started");
 
+    if (context.isCancelled()) {
+      // Context is already cancelled, cancel immediately and notify the observer of cancellation
+      // via callback on the executor
+      cancel();
+      callExecutor.execute(new ContextRunnable(context) {
+        @Override
+        public void runInContext() {
+          // Must call onHeaders per API contract
```

Comment:
Is that our contract? I thought you could just call `onClose()`. We do require headers before any messages, though.